### PR TITLE
Default plan selector - change to PF typeahead select component

### PR DESCRIPTION
**Pull request by srambach**
**What this PR does / why we need it**:

Changes the default plan selector in Application Plans to be a React/Patternfly component

select

**Which issue(s) this PR fixes** 

[APPDUX-764: [3scale][2.11][LO-prio] Improve select default Application plan](https://issues.redhat.com/browse/APPDUX-764)

**Verification steps** 

Select Product > Applications > Application Plans > Default Plan card
(url: apiconfig/services/:id/application_plans)
Marvel mockup: https://marvelapp.com/prototype/6g85cf6/screen/76056565

### CODE REVIEW

**Comment by cntlsn** [comment]
@srambach @josemigallas I'm a bit confused about what UI component this PR is actually changing, the linked mockup doesn't match with the Jira issue and the verification steps.
Is this about the 'Default Application plan' select or the 'Change Application plan' select?

**Comment by josemigallas** [comment]
> @srambach @josemigallas I'm a bit confused about what UI component this PR is actually changing, the linked mockup doesn't match with the Jira issue and the verification steps.
> Is this about the 'Default Application plan' select or the 'Change Application plan' select? Or both?

You're right, we mixed up both Jiras. It should be OK now!

**Review by cntlsn — APPROVED** [review]
It might need a round of microcopy edits, let's wait for @porueesq to review it.
Component looks good!

File: app/controllers/api/plans_base_controller.rb
Comment (guicassolato) [inline app/controllers/api/plans_base_controller.rb:122]: I see some DRY opportunity here when comparing to `#generic_masterize_plan`

File: app/controllers/api/plans_base_controller.rb
Comment (guicassolato) [inline app/controllers/api/plans_base_controller.rb:115]: Are we intending to use this method in another controller the inherits from `Api::PlansBaseController`?

File: app/controllers/api/plans_base_controller.rb
Comment (josemigallas) [reply app/controllers/api/plans_base_controller.rb:122]: `generic_masterize_plan` will be removed in a following PR (right after this one is merged), after all of the "old" selects have been replaced with this one. At which point there will be only one `#masterize_plan` method, that's why I don't think refactoring is worth it BUT any suggestions welcome.

File: app/controllers/api/plans_base_controller.rb
Comment (josemigallas) [reply app/controllers/api/plans_base_controller.rb:115]: Not that I am aware of.

**Review by porueesq — APPROVED** [review]
LGTM :+1:

File: app/views/master/providers/plans/_widget.html.slim
Comment (didierofrivia) [inline app/views/master/providers/plans/_widget.html.slim:23]: Plan PEPE?

File: app/javascript/src/utilities/patternfly-utils.jsx
Comment (didierofrivia) [inline app/javascript/src/utilities/patternfly-utils.jsx:19]: Would be nice to have a factory instead of a class

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (didierofrivia) [inline app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:59]: I haven't seen any tests asserting the different responses and how the UI would behave

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (didierofrivia) [inline app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:34]: A bit dangerous to treat it like a string. Maybe we could make path of type URL?

File: app/javascript/src/Common/components/Spinner.scss
Comment (didierofrivia) [inline app/javascript/src/Common/components/Spinner.scss:1]: This means they haven't written tests for the Spinner yet, right?

File: app/controllers/api/plans_base_controller.rb
Comment (didierofrivia) [inline app/controllers/api/plans_base_controller.rb:31]: wouldn't it raise with the `.presence` ?

File: app/javascript/packs/default_plan_selector.js
Comment (damianpm) [inline app/javascript/packs/default_plan_selector.js:5]: Not important, but I'd omit new ine

File: app/javascript/src/Common/components/Spinner.jsx
Comment (damianpm) [inline app/javascript/src/Common/components/Spinner.jsx:13]: Is `toString()` needed if size is already a string?
Why not using default value for `className`:
```suggestion
const Spinner = ({ size = 'md', className = '' }: Props) => {
  const classParameters = `pf-c-spinner pf-m-${size} ${className}`
```

File: app/javascript/src/Common/components/Spinner.jsx
Comment (damianpm) [inline app/javascript/src/Common/components/Spinner.jsx:24]: Add new line at end of files, or our linter will complain

File: app/javascript/src/Plans/components/DefaultPlanSelect.jsx
Comment (damianpm) [inline app/javascript/src/Plans/components/DefaultPlanSelect.jsx:20]: Can you add a note explaining why? It would be good when working with this in the future

File: app/javascript/src/Plans/components/DefaultPlanSelect.jsx
Comment (damianpm) [inline app/javascript/src/Plans/components/DefaultPlanSelect.jsx:15]: Other arrow functions with a single argument don't have the `()`, like line 28

File: app/javascript/src/Plans/components/DefaultPlanSelect.jsx
Comment (damianpm) [inline app/javascript/src/Plans/components/DefaultPlanSelect.jsx:36]:
```suggestion
  const onClear = () => setSelection(null)
```

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (damianpm) [inline app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:18]: I'd rather omit empty lines, and group all imports as a block

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (damianpm) [inline app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:74]: I'd extract this into a const, instead of filtering inline

File: app/views/api/plans/_default_plan.html.slim
Comment (damianpm) [inline app/views/api/plans/_default_plan.html.slim:1]: WDYM with deprecated? The file should be removed?

Comment (josemigallas) [reply app/views/api/plans/_default_plan.html.slim:1]: Yes, all plan selectors will be replaced by this one and this file removed (in another PR or PRs).

**Comment by damianpm** [comment]
@josemigallas have you tried this locally?
Maybe we can create a feature branch, merge there, deploy to preview and test it, before merging to master.

File: app/views/master/providers/plans/_widget.html.slim
Comment (josemigallas) [reply app/views/master/providers/plans/_widget.html.slim:23]: Oops

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (josemigallas) [reply app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:34]: In this case path is "guaranteed" to be a string but in any case, URL can't be used for paths but for a full URL, which we don't know.

File: app/javascript/src/Plans/components/DefaultPlanSelect.jsx
Comment (josemigallas) [reply app/javascript/src/Plans/components/DefaultPlanSelect.jsx:20]: It's hard to explain, it's some kind of bug and it complains about a prop being `number` instead of `string`, but the type is defined as `number | string`. Not sure how to explain that 😟

File: app/javascript/src/Common/components/Spinner.jsx
Comment (josemigallas) [reply app/javascript/src/Common/components/Spinner.jsx:24]: But there is one 😬

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (josemigallas) [reply app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:18]: Everything for an approval 🙏 🤣

File: app/controllers/api/plans_base_controller.rb
Comment (josemigallas) [reply app/controllers/api/plans_base_controller.rb:31]: @guicassolato

File: app/javascript/src/Plans/components/DefaultPlanSelect.jsx
Comment (josemigallas) [reply app/javascript/src/Plans/components/DefaultPlanSelect.jsx:20]: OK I came up with something in the end

File: app/controllers/api/plans_base_controller.rb
Comment (hallelujah) [reply app/controllers/api/plans_base_controller.rb:31]: No basically `#presence` works with `nil`

It does something like:

```ruby
def presence
  present? ? self : nil
end

### e.g

"".presence # => nil
nil.presence # => nil
1.presence # => 1
Object.new.presence # => #<Object:0x00007fcc731ffd78>
```

File: app/controllers/api/plans_base_controller.rb
Comment (hallelujah) [inline app/controllers/api/plans_base_controller.rb:34]: You wanted to do probably this:

```suggestion
  def resource(id = params[:id])
    return unless id.present?
    collection.readonly(false).find(id)
  end
```

Because in your version if you call `resource("")` it will still execute `collection.readonly(false).find("")`

File: app/controllers/api/plans_base_controller.rb
Comment (hallelujah) [inline app/controllers/api/plans_base_controller.rb:117]: Basically if it is the same plan passed then you would unset it? That looks strange to me.

Comment (josemigallas) [reply app/controllers/api/plans_base_controller.rb:117]: @guicassolato

File: app/controllers/api/plans_base_controller.rb
Comment (josemigallas) [reply app/controllers/api/plans_base_controller.rb:34]: @guicassolato

File: app/controllers/api/plans_base_controller.rb
Comment (guicassolato) [reply app/controllers/api/plans_base_controller.rb:117]: I haven's changed the logic of it.

**Comment by didierofrivia** [comment]
The JS part looks good... however, there are still some cukes and units failing.

**Review by damianpm — APPROVED** [review]
JS part LGTM, let's wait for a Ruby review

File: app/helpers/api/plans_helper.rb
Comment (hallelujah) [inline app/helpers/api/plans_helper.rb:18]: Probably we should put this code into the controller, assigning a variable.

But it is OK for now

**Review by hallelujah — APPROVED** [review]
Approved. New PR welcome to improve it!

Thank you for the job 🚀

File: app/javascript/src/utilities/alert.js
Comment (hallelujah) [inline app/javascript/src/utilities/alert.js:5]: <3

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (hallelujah) [inline app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:13]: Nice one! I like this, at least we can change it to `import * as alert from 'jquery'` later

**Review by hallelujah — APPROVED** [review]

File: app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx
Comment (josemigallas) [reply app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:13]: My idea is to eventually remove "flash" from the UI and start using Patternfly's `Alert`. This way we'll only have to change the implementation of `utilities/alert`.

Comment (hallelujah) [reply app/javascript/src/Plans/components/DefaultPlanSelectCard.jsx:13]: Yes that is the way to go, please remove assets pipeline from the code, it should be handled by webpack

**Pull request merged**
